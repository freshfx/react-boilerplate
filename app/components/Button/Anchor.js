import styled from '@emotion/styled'

import buttonStyles from './buttonStyles'

const Anchor = styled.a`
  ${buttonStyles};
`

export default Anchor
