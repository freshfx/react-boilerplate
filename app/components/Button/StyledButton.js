import styled from '@emotion/styled'

import buttonStyles from './buttonStyles'

const StyledButton = styled.button`
  ${buttonStyles};
`

export default StyledButton
