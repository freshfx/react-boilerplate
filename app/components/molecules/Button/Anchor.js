import styled from '@emotion/styled'

import buttonStyles from 'components/atoms/Button/buttonStyles'

const Anchor = styled.a`
  ${buttonStyles};
`

export default Anchor
