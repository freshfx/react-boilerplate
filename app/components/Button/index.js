/**
 *
 * Button.js
 *
 * A common button, if you pass it a prop "route" it'll render a link to a react-router route
 * otherwise it'll render a link with an onclick
 */

import React, {Children} from 'react'
import PropTypes from 'prop-types'

import Anchor from './Anchor'
import StyledButton from './StyledButton'
import Wrapper from './Wrapper'

function Button(props) {
  // Render an anchor tag
  let button = (
    <Anchor href={props.href} onClick={props.onClick}>
      {Children.toArray(props.children)}
    </Anchor>
  )

  // If the Button has a handleRoute prop, we want to render a button
  if (props.handleRoute) {
    button =
      <StyledButton onClick={props.handleRoute}>
        {Children.toArray(props.children)}
      </StyledButton>
  }

  return <Wrapper>{button}</Wrapper>
}

Button.propTypes = {
  children: PropTypes.node.isRequired,
  handleRoute: PropTypes.func,
  href: PropTypes.string,
  onClick: PropTypes.func
}

Button.defaultProps = {
  handleRoute: null,
  href: '',
  onClick: null
}

export default Button
