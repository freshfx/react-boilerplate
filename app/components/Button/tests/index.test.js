import React from 'react'
import {render, fireEvent} from '@testing-library/react'

import Button from '../index'

const handleRoute = () => {}

const renderComponent = (props = {}) => render(
  <Button href="http://mxstbr.com" {...props}>
    <h1>Test</h1>
  </Button>
)

describe('Button', () => {
  it('should render an anchor tag if no route is specified', () => {
    const {container} = renderComponent()
    expect(container.querySelector('a')).not.toBeNull()
  })

  it('should render a button tag to change route if the handleRoute prop is specified', () => {
    const {container} = renderComponent({handleRoute})
    expect(container.querySelector('button')).toBeDefined()
  })

  it('should have children', () => {
    const {container} = renderComponent()
    expect(container.querySelector('a').children).toHaveLength(1)
  })

  it('should handle click events', () => {
    const onClickSpy = jest.fn()
    const {container} = renderComponent({onClick: onClickSpy})
    fireEvent.click(container.querySelector('a'))
    expect(onClickSpy).toHaveBeenCalled()
  })

  it('should have a className attribute', () => {
    const {container} = renderComponent()
    expect(container.querySelector('a').hasAttribute('class')).toBe(true)
  })

  it('should not adopt a type attribute when rendering an anchor tag', () => {
    const type = 'text/html'
    const {container} = renderComponent({type})
    expect(container.querySelector(`a[type="${type}"]`)).toBeNull()
  })

  it('should not adopt a type attribute when rendering a button', () => {
    const type = 'submit'
    const {container} = renderComponent({handleRoute, type})
    expect(container.querySelector('button').getAttribute('type')).toBeNull()
  })
})
