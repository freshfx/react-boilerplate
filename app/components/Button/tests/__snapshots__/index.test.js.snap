// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should render a button tag to change route if the handleRoute prop is specified 1`] = `
.emotion-2 {
  width: 100%;
  text-align: center;
  margin: 4em 0;
}

.emotion-0 {
  display: inline-block;
  box-sizing: border-box;
  padding: 0.25em 2em;
  -webkit-text-decoration: none;
  text-decoration: none;
  border-radius: 4px;
  -webkit-font-smoothing: antialiased;
  -webkit-touch-callout: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: pointer;
  outline: 0;
  font-family: 'Helvetica Neue',Helvetica,Arial,sans-serif;
  font-weight: bold;
  font-size: 16px;
  border: 2px solid #41addd;
  color: #41addd;
}

.emotion-0:active {
  background: #41addd;
  color: #fff;
}

<div
  class="emotion-2 emotion-3"
>
  <button
    class="emotion-0 emotion-1"
  >
    <h1>
      Test
    </h1>
  </button>
</div>
`;

exports[`Button should render an anchor tag if no route is specified 1`] = `
.emotion-2 {
  width: 100%;
  text-align: center;
  margin: 4em 0;
}

.emotion-0 {
  display: inline-block;
  box-sizing: border-box;
  padding: 0.25em 2em;
  -webkit-text-decoration: none;
  text-decoration: none;
  border-radius: 4px;
  -webkit-font-smoothing: antialiased;
  -webkit-touch-callout: none;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  cursor: pointer;
  outline: 0;
  font-family: 'Helvetica Neue',Helvetica,Arial,sans-serif;
  font-weight: bold;
  font-size: 16px;
  border: 2px solid #41addd;
  color: #41addd;
}

.emotion-0:active {
  background: #41addd;
  color: #fff;
}

<div
  class="emotion-2 emotion-3"
>
  <a
    class="emotion-0 emotion-1"
    href="http://mxstbr.com"
  >
    <h1>
      Test
    </h1>
  </a>
</div>
`;
